### DISCUSSION TITLE
Test cases revealing strange drone behavior

### DISCUSSION THREAD

**Discussion by dgumenyuk** [skhatiri/UAV-Testing-Competition#11] [Test Generation] [answered]
I am attaching specifications as well as logs of some cases produced with out test generator that cause some strange drone behavior.
I think it will be interesting to investigate whether it is due to invalid test cases, bugs in the drone controller or simulation environment bugs. [logs](https://drive.google.com/file/d/1IQM5SzmsNqkdScakWVLrLrwL7QcruiuC/view?usp=sharing)
test 1:
![test_39](https://github.com/skhatiri/UAV-Testing-Competition/assets/17391422/3fe9ca35-223e-4358-827d-05afa5e193a1)
test 2:
![test_34](https://github.com/skhatiri/UAV-Testing-Competition/assets/17391422/765b7457-94be-4d72-b094-3116c80112fd)
test 3:
![test_27](https://github.com/skhatiri/UAV-Testing-Competition/assets/17391422/8762e926-eb10-4809-9070-73634af64784)
test 4:
![test_4](https://github.com/skhatiri/UAV-Testing-Competition/assets/17391422/480994cd-fc3c-41b7-92a1-d69298c5e9fe)
test 5:
![test_2](https://github.com/skhatiri/UAV-Testing-Competition/assets/17391422/fbece4b7-8500-480f-beaf-e49791834016)

Mission file is "mission1" from the case studies folder.

**Comment by skhatiri** [comment, marked as answer]
Hi @dgumenyuk, Thanks for the interesting questions and plots! 

Such test cases are indeed crashing scenarios, with minimum distance = 0. 
There are some potential reasons for this strange behavior seen in the plots
- These plots are extracted from *local position* data estimated at runtime by the drone. Sometimes, a physical crash of the drone can lead these estimations to go bad! and include noisy data for a few ms, which is more obvious in the time-based plots on the left. So, in most of such scenarios, the drone did not actually fly that strange path, it is just wrongly estimating its current position after the crash. 
   - We have tried to use a more reliable ground truth position directly from the simulator instead of this estimation, but it was not more reliable in many cases. in the future, we may be able to fix this. 
- In order to mitigate such physical crash effects in the test cases and let the drone fly till the end to make things simpler in the competition, we implemented *collision-free borders* for the obstacles: 
This is more seen in your test 4 plot which looks like the drone is flying inside the obstacle (which is actually true!):
![289989791-480994cd-fc3c-41b7-92a1-d69298c5e9fe](https://github.com/skhatiri/UAV-Testing-Competition/assets/16767722/6aeeb5b8-0727-4749-a7a3-025de8412f42)
Since this [commit](https://github.com/skhatiri/Aerialist/commit/f76b46cb262cf848d71f7c93747fa3f7e84551f1) the physical size of the obstacle is about 80% of its visual size. if you set a length and width of 10 m, the obstacle is visually seen as 10\*10m, which is the only important factor for the vision-based obstacle avoidance system, but it has a physical size of 8\*8m, giving it a 1m collision-free border from each side (this does not affect the height). So, if the drone touches the borders, it will not physically crash to it and can still be able to continue its path. Of course,  the visual size (that you provide) is used for all the distance calculations and plots, and such scenarios count as a *hard fail* in evaluations. 

Such decisions are mostly taken to simplify the pipeline and competition rules. I hope it is more clear now.
